### PR TITLE
CSV Support

### PR DESCRIPTION
**Pull request by ehsantn**

### CODE REVIEW

**Comment by fschlimb** [comment]
@ehsantn: changes are mostly yours, mine are mostly what happens in _csv.cpp/h. the changes in other files are simple cut&paste to share functionality through header files and marking them 'static' (as agreed). I will add comments where more than that happens.

File: hpat/_hpat_common.h [hunk 1]
```diff
@@ -1,6 +1,11 @@
 #ifndef HPAT_COMMON_H_
 #define HPAT_COMMON_H_
 
+#if defined(__GNUC__)
+#define __UNUSED__ __attribute__((unused))
+#else
+#define __UNUSED__
+#endif
```
Comment (fschlimb) [inline hpat/_hpat_common.h]: adding a macro to mark possible unused objects

File: hpat/_import_py.h [hunk 2]
```diff
@@ -0,0 +1,27 @@
...
+    CHECK(func, "getting symbol from module failed");
+
+    Py_XDECREF(mod);
+    PyGILState_Release(gilstate);
+
+    return func;
+#undef CHECK
+}
+
+#endif // _IMPORT_PY_INCLUDED
```
Comment (fschlimb) [inline hpat/_import_py.h]: Generic way to import python-exported C funcs

File: hpat/_io.cpp [hunk 3]
```diff
@@ -82,6 +83,9 @@ PyMODINIT_FUNC PyInit_hio(void) {
                             PyLong_FromVoidPtr((void*)(&file_read_parallel)));
     PyObject_SetAttrString(m, "file_write_parallel",
                             PyLong_FromVoidPtr((void*)(&file_write_parallel)));
+
+    PyInit_csv(m);
+
```
Comment (fschlimb) [inline hpat/_io.cpp]: Call initializer for csv stuff

File: hpat/_meminfo.h [hunk 4]
```diff
@@ -0,0 +1,30 @@
...
+    PyGILState_Release(gilstate);
+    return fnptr;
+#undef CHECK
+}
+
+typedef void (*MemInfo_release_type)(void*);
+typedef MemInfo* (*MemInfo_alloc_aligned_type)(size_t size, unsigned align);
+typedef void* (*MemInfo_data_type)(MemInfo* mi);
+
+#endif
```
Comment (fschlimb) [inline hpat/_meminfo.h]: file unused, could be removed.

**Comment by ehsantn** [comment]
OK, I'll take a quick look and squash/merge.

**Pull request merged**